### ISSUE TITLE
_.union doesn't work with arrays of objects

### ISSUE THREAD

**Issue by wilhen01** [jashkenas/underscore#2311]
_.union will always produce duplicates when passed arrays of objects. 

e.g. `_.union( [ { a:1 } ], [ { a:1 } ])` will return `[ { a:1 }, { a:1 } ]`

Perversely, underscore's own isEqual function will tell you that the objects in question are equal. Maybe we could have a flag/option which dictates the equality comparison to use, or the option to pass in a comparator?

**Comment by michaelficarra** [comment]
I'm surprised it doesn't already accept the comparison function. :+1:

**Comment by wilhen01** [comment]
It's worth pointing out that this applies to all the other array computation functions like difference, intersection, unique, etc.

**Comment by amiral84** [comment]
Wouldn't it be easier if we would have an option that compares equality based on boolean parameter value that could be passed to _.union() function? If it is true then it would automatically compare all objects in that array.

E.g. <code>_.union([1, 2, 3, 10, [{a:1}, {a:1}]], true)</code>, would output <code>[1,2,3,10, {a:1}]</code>

**Comment by michaelficarra** [comment]
@amiral84 No. That is unrelated. If you want that behaviour, compose union with flatten.

**Comment by amiral84** [comment]
@michaelficarra Then did I miss the point of this topic? :D

**Comment by michaelficarra** [comment]
@amiral84 It appears so. The feature request is explained fully and succinctly in the first comment.

**Comment by lucasfanning** [comment]
because ({a:1} == {a:1}) === false;

**Comment by lucasfanning** [comment]
Because {a:1} != {a:1}

**Comment by dperrymorrow** [comment]
the underlying problem seems to be in ```_.unique``` as ```_.union``` is merely a wrapper function for unique and flatten.

```javascript
_.union = restArgs(function(arrays) {
  return _.uniq(flatten(arrays, true, true));
});
```

**Comment by jdalton** [comment]
This thread inspired me to add `_.intersectionWith`, `_.differenceWith`, `_.unionWith`, and `_.uniqWith` to handle comparison customization.

```js
_.uniqWith([ { 'a': 1, 'b': 2 }, { 'a': 1, 'b': 3 }, { 'a': 1, 'b': 2 } ], _.isEqual);
// => [{ 'a': 1,'b': 2 }, { 'a': 1,'b': 3 }]
```

**Comment by dperrymorrow** [comment]
or something along the lines of ```_.isCollection``` to determine if you are dealing with a collection. If dealing with a collection, then the comparisons should use ```_.isEqual``` instead of  ```===``` which does no good in the case of a collection.

**Comment by jdalton** [comment]
>  use ```_.isEqual``` instead of  ```===``` which does no good in the case of a collection.

Dynamic switching sounds like a bad idea. JS uses `===` or `SameValueZero` for many things. If there's a need to go outside those comparisons something like `_.uniqWith` would do.

**Comment by wilhen01** [comment]
Thanks for this @jdalton, the _opWith functions you mention are absolutely perfect for what I'm trying to achieve. Any idea when they'll be available via release?

**Comment by dperrymorrow** [comment]
@jdalton good point on the comparisons, but wouldn't you usually use a key for unique on a collection rather than forcing Underscore to detect the entire difference between the objects?

Wouldn't the following solve @wilhen01 's request _(albeit more verbose than desired)_

```javascript
_.chain([{ a: 1 }]).union( [{a: 1}]).unique('a').value();
//=> [{a: 1}]
```

**Comment by jdalton** [comment]
> Wouldn't the following solve @wilhen01 's request (albeit more verbose than desired)

`_.uniq` already supports that.

**Comment by dperrymorrow** [comment]
right, thats my point, the above code works currently as posted.
could't you just call uniq / unique with a key on the result of the union?

**Comment by jdalton** [comment]
@dperrymorrow Think just an inch outside that example [and add another property](https://github.com/jashkenas/underscore/issues/2311#issuecomment-159739542).

**Comment by dperrymorrow** [comment]
ok, gotcha, sorry... Im not trying to be belligerent, just wanted to totally understand the problem. Id love to submit a pull request on the ```_.uniqWith``` function.

**Comment by jdalton** [comment]
No worries, that would be rad.

**Comment by jashkenas** [comment]
> _.intersectionWith, _.differenceWith, _.unionWith, and _.uniqWith

Wouldn't it be a nicer API to just allow the comparison function to be optionally passed as the final argument, instead of minting four new functions?

**Comment by jdalton** [comment]
@jashkenas 
> Wouldn't it be a nicer API to just allow the comparison function to be optionally passed as the final argument, instead of minting four new functions?

Yep, it could be done but there's complications because methods like `_.uniq` already support passing an iteratee and [are already heavily overloaded](https://github.com/jashkenas/underscore/blob/cf3f7a7276055897a7434acdedd1089a5ebe04f4/underscore.js#L536) with support for binary search flags and context params too . This would mean introducing arity sniffing which feels too clever for this situation. This would also complicate future modularization efforts because it bundles in lots of optional functionality into a single point when the implementations could be simplified and split into separate methods.

**Comment by jashkenas** [comment]
Right, a totally unfortunate design issue. But making new functions just to allow a comparator doesn't feel like the right solution either.

**Comment by dperrymorrow** [comment]
ok so additional comparison function parameters are the way to go here then? 
If so I can update my pull request.

the only tricky part I forsee is making the parameter parsing a little more hairy as @jdalton mentioned above.

```javascript
_.uniq = _.unique = function(array, isSorted, iteratee, context) {
  if (!_.isBoolean(isSorted)) {
    context = iteratee;
    iteratee = isSorted;
    isSorted = false;
  }
//...
```
Perhaps adding an additional check to see if isSorted ```_.isFunction``` then treat it as the comparator.

**Comment by jdalton** [comment]
> But making new functions just to allow a comparator doesn't feel like the right solution either.

It may be the best option for a bad situation. I've gone on a kick of splitting out overloaded functionality. Though it increases the API surface it allows for simpler implementations. In the case of `uniq` though I still use a shared base function.

**Comment by dperrymorrow** [comment]
have updated this pull request #2368  thanks.

**Comment by megawac** [comment]
I'm :+1: for `uniqBy` or `uniqWith`. I would be completely against overloading `uniq` further (as #2368 currently is proposed)

**Comment by michaelficarra** [comment]
:+1: @megawac, `uniqBy`.

**Comment by jdalton** [comment]
Fwiw lodash will be using `uniqBy` as the split out form of `_.uniq(array, iteratee)` and `_.uniqWith` as the form to allow comparator customization.

**Comment by megawac** [comment]
Yeah, on second thought `uniqWith` is a better name

**Comment by dperrymorrow** [comment]
should I pull request on LowDash with the separate method then? 
I thought the two projects merged, am i mistaken?

**Comment by jdalton** [comment]
@dperrymorrow 
> should I pull request on Lodash with the separate method then

No need, they're already in lodash's edge master branch.

> I thought the two projects merged, am i mistaken?

Not yet. Lodash v4 (due out in Jan) proofs out some of the ideas of the merge though.

**Comment by Pavnii** [comment]
@jdalton I am a newbie to underscore.js. Can you please elaborate a liitle on any of these implementation:_.intersectionWith, _.differenceWith, _.unionWith, _.uniqWith.

**Comment by jdalton** [comment]
@Pavnii 
Sure. You can check out [lodash/npm/_baseUniq](https://github.com/lodash/lodash/blob/4.3.0-npm/_baseUniq.js#L61). If a `comparator` is passed it uses `arrayIncludesWith` helper to do the check instead of `arrayIncludes` (Underscore's `contains`).

**Comment by Pavnii** [comment]
@jdalton That helps me.
